File: packages/postgres-db/src/create-db-user-handler.ts
import type { CloudFormationCustomResourceEvent } from 'aws-lambda';
import { randomBytes } from 'node:crypto';
import { createPool, type PoolConnection } from 'mariadb';
import { getDatabaseSecret } from './utils.js';

type OnEventResult = {
  PhysicalResourceId: string;
  Data?: Record<string, unknown>;
};

const physicalResourceIdPrefix = 'db-user:';

const resolveDbUser = (physicalResourceId?: string): string =>
  physicalResourceId?.startsWith(physicalResourceIdPrefix)
    ? physicalResourceId.slice(physicalResourceIdPrefix.length)
    : `db_${randomBytes(8).toString('hex')}`;

const ensureDatabaseUser = async (dbUser: string): Promise<void> => {
  const { dbname, username, password, host, port } = await getDatabaseSecret();
  const pool = createPool({
    host,
    port,
    database: dbname,
    user: username,
    password,
    ssl: {
      rejectUnauthorized: true,
    },
    connectionLimit: 1,
    multipleStatements: true,
  });

  const quotedDbName = pool.escapeId(dbname);
  const quotedUser = pool.escape(dbUser);
  const quotedHost = pool.escape('%');

  let connection: PoolConnection | undefined;

  try {
    connection = await pool.getConnection();
    await connection.query(
      [
        `CREATE USER IF NOT EXISTS ${quotedUser}@${quotedHost} IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS'`,
        `ALTER USER ${quotedUser}@${quotedHost} IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS' REQUIRE SSL`,
        `GRANT ALL PRIVILEGES ON ${quotedDbName}.* TO ${quotedUser}@${quotedHost}`,
      ].join(';\n'),
    );
  } finally {
    await connection?.release();
    await pool.end();
  }
};

export const handler = async (
  event: CloudFormationCustomResourceEvent,
): Promise<OnEventResult> => {
  const dbUser = resolveDbUser(
    'PhysicalResourceId' in event ? event.PhysicalResourceId : undefined,
  );

  if (event.RequestType !== 'Delete') {
    await ensureDatabaseUser(dbUser);
  }

  return {
    PhysicalResourceId: `${physicalResourceIdPrefix}${dbUser}`,
    Data: {
      dbUser,
    },
  };
};

File: packages/my_sql_db/aws_nx_poc_my_sql_db/create_db_user_handler.py
import asyncio
import secrets
import ssl
from typing import Any

import asyncmy

from .utils import get_database_secret, with_connection_retry


async def _ensure_database_user(db_user: str) -> None:
    secret = get_database_secret()
    conn = await asyncmy.connect(
        host=secret["host"],
        port=int(secret["port"]),
        db=secret["dbname"],
        user=secret["username"],
        password=secret["password"],
        ssl=ssl.create_default_context(),
    )
    try:
        async with conn.cursor() as cursor:
            await cursor.execute(
                "CREATE USER IF NOT EXISTS %s@'%%' IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS'",
                (db_user,),
            )
            await cursor.execute(
                "ALTER USER %s@'%%' IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS' REQUIRE SSL",
                (db_user,),
            )
            await cursor.execute(
                f"GRANT SELECT, INSERT, UPDATE, DELETE ON `{secret['dbname']}`.* TO %s@'%%'",
                (db_user,),
            )
        await conn.commit()
    finally:
        conn.close()


_PHYSICAL_RESOURCE_ID_PREFIX = "db-user:"


def _resolve_db_user(physical_resource_id: str | None) -> str:
    if physical_resource_id and physical_resource_id.startswith(_PHYSICAL_RESOURCE_ID_PREFIX):
        return physical_resource_id[len(_PHYSICAL_RESOURCE_ID_PREFIX) :]
    return f"db_{secrets.token_hex(8)}"


def handler(event: dict[str, Any], _context: Any) -> dict[str, Any]:
    db_user = _resolve_db_user(event.get("PhysicalResourceId"))

    if event.get("RequestType") != "Delete":
        with_connection_retry(lambda: asyncio.run(_ensure_database_user(db_user)))

    return {
        "PhysicalResourceId": f"{_PHYSICAL_RESOURCE_ID_PREFIX}{db_user}",
        "Data": {"dbUser": db_user},
    }
